### PR TITLE
Support publication of Courses with a credit seat

### PR DESCRIPTION
**Pull request by rlucioni**
Creates or updates a CreditCourse object on the LMS when publishing a Course with a credit seat. XCOM-585.

@clintonb and/or @jimabramson, please review.

### CODE REVIEW

File: ecommerce/courses/publishers.py [hunk 1]
```diff
@@ -50,6 +87,36 @@ def publish(self, course):
...
         verification_deadline = self.get_course_verification_deadline(course)
         modes = [self.serialize_seat_for_commerce_api(seat) for seat in course.seat_products]
+
+        has_credit = 'credit' in [mode['name'] for mode in modes]
+        if has_credit:
+            if access_token is not None:
+                try:
+                    response = self._publish_creditcourse(course_id, access_token)
+
+                    if response.status_code == 400:
```
Comment (clintonb) [inline ecommerce/courses/publishers.py]: Move this logic into `_publish_creditcourse`.

File: ecommerce/courses/tests/test_publishers.py [hunk 2]
```diff
@@ -37,6 +38,31 @@ def _mock_commerce_api(self, status, body=None):
         httpretty.register_uri(httpretty.PUT, url, status=status, body=json.dumps(body),
                                content_type=JSON)
 
+    def _mock_credit_api(self, creation_status, update_status):
+        self.assertTrue(httpretty.is_enabled, 'httpretty must be enabled to mock Credit API calls.')
+
+        url = '{host}/{path}'.format(
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publishers.py]: Use `get_lms_url`.

File: ecommerce/courses/publishers.py [hunk 3]
```diff
@@ -30,14 +32,49 @@ def serialize_seat_for_commerce_api(self, seat):
...
+        }
+
+        kwargs = {
+            'url': url,
+            'data': json.dumps(data),
+            'headers': headers,
+            'timeout': self.timeout
+        }
+        if update:
+            kwargs['url'] += course_id
```
Comment (clintonb) [inline ecommerce/courses/publishers.py]: Missing the trailing slash.

File: ecommerce/courses/tests/test_publishers.py [hunk 4]
```diff
@@ -139,3 +165,59 @@ def test_serialize_seat_for_commerce_api_with_professional(self, is_verified, ex
...
+
+    @httpretty.activate
+    @ddt.data(
+        ('access_token', 201, None, 201, True),
+        ('access_token', 400, 200, 200, True),
+        ('access_token', 400, 418, 200, False),
+        (None, None, None, None, False)
+    )
+    @ddt.unpack
+    def test_credit_publication(self, access_token, creation_status, update_status, commerce_status, success):
```
Comment (clintonb) [inline ecommerce/courses/tests/test_publishers.py]: There is a lot of magic going on here. This test suite would be easier to understand if this test was broken into multiple tests that validate the explicit error states.

File: ecommerce/extensions/api/v2/views.py [hunk 5]
```diff
@@ -538,6 +539,11 @@ class AtomicPublicationView(generics.CreateAPIView, generics.UpdateAPIView):
     permission_classes = (IsAuthenticated, IsAdminUser,)
     serializer_class = serializers.AtomicPublicationSerializer
 
+    def get_serializer_context(self):
+        context = super(AtomicPublicationView, self).get_serializer_context()
+        context['access_token'] = getattr(self.request.user, 'access_token', None)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views.py]: Why not use `request.user.access_token`, which returns `None` if no token exists?

Comment (rlucioni) [inline ecommerce/extensions/api/v2/views.py]: I didn't realize the property returns `None` if no token exists. Updated.

File: ecommerce/courses/publishers.py (hunk 1 above)
Comment (rlucioni) [inline ecommerce/courses/publishers.py]: Moved.

File: ecommerce/courses/tests/test_publishers.py (hunk 2 above)
Comment (rlucioni) [inline ecommerce/courses/tests/test_publishers.py]: Updated.

File: ecommerce/courses/publishers.py (hunk 3 above)
Comment (rlucioni) [inline ecommerce/courses/publishers.py]: Added.

File: ecommerce/courses/publishers.py [hunk 6]
```diff
@@ -30,14 +32,49 @@ def serialize_seat_for_commerce_api(self, seat):
...
-        Uses the Commerce API to publish course modes, prices, and SKUs to LMS.
+        Uses the Commerce API to publish course modes, prices, and SKUs to LMS. Uses
+        CreditCourse API endpoints to publish CreditCourse data to LMS when necessary.
 
-        Args:
+        Arguments:
             course (Course): Course to be published.
 
+        Keyword Arguments:
+            access_token (str): Access token used when publishing CreditCourse data to the LMS.
```
Comment (jimabramson) [inline ecommerce/courses/publishers.py:75]: why is access_token required for CreditCourse publishing but not for other types / modes?

Comment (rlucioni) [inline ecommerce/courses/publishers.py:75]: The Credit API uses OAuth2 for authentication, not an API key, like the Commerce API. Only publication of Courses with a credit seat requires making calls to the Credit API.

File: ecommerce/courses/tests/test_publishers.py (hunk 4 above)
Comment (rlucioni) [inline ecommerce/courses/tests/test_publishers.py]: @clintonb broken up.

File: ecommerce/courses/publishers.py (hunk 6 above)
Comment (clintonb) [inline ecommerce/courses/publishers.py:75]: Correct. I have a preference toward OAuth2 and any other authentication scheme that explicitly associates a request with a user. Our server-to-server authentication does not provide this, meaning we have to create minor hacks to deal with permissions.

**Comment by clintonb** [comment]
:+1:

**Comment by jimabramson** [comment]
:+1:

**Pull request merged**